src/keyboard_drag_strategy.ts

```diff
@@ -289,4 +283,77 @@ export class KeyboardDragStrategy extends dragging.BlockDragStrategy {
   override shouldHealStack(e: PointerEvent | undefined): boolean {
     return true;
   }
+
+  /**
+   * Get the previous node in the tree, with loopback to the last
+   * stack on the workspace if needed.
+   *
+   * @param start Where to start traversal.
+   * @param cursor The workspace's cursor
+   * @returns The previous node, or null if there were no valid nodes
+   *     on the workspace.
+   */
+  private getPreviousNode(start: ASTNode, cursor: LineCursor) {
+    let potential: ASTNode | null = start;
+    potential = cursor.getPreviousNode(potential, (node) => {
+      // @ts-expect-error isConnectionType is private.
+      return node && ASTNode.isConnectionType(node.getType());
+    });
+    if (!potential) {
+      // Loop back to last block if it exists.
+      // @ts-expect-error workspace is private
+      const topBlocks = this.workspace.getTopBlocks(true);
+      if (!topBlocks.length) return null;
+
+      // Find the last stack.
+      const lastTopBlockNode = ASTNode.createStackNode(
+        topBlocks[topBlocks.length - 1],
+      );
+      let prevNode = lastTopBlockNode;
+      let nextNode: ASTNode | null = lastTopBlockNode;
+      // Iterate until you fall off the end of the stack.
+      while (nextNode) {
+        prevNode = nextNode;
+        nextNode = cursor.getNextNode(prevNode, (node) => {
```

Comment:
I could be wrong here, but if the very last thing in this stack is a compatible connection, would this code back up too far? On this line prevNode would be set to the connection node, then when you resume searching, you'd back up an extra step?

It seems like you should have an extra step to check if the prevNode is a compatible connection already, if it is return, else continue searching. But I could be misunderstanding something. otherwise, same applies to the `getNextNode` logic.